### PR TITLE
feat(authorization): refactor role_assignment resource and implement folder assignment

### PR DESCRIPTION
**Pull request by h3adex**
## Description

This PR refactors `stackit_authorization_{project,folder,organization}_role_assignment` resources to align with our coding standards and adds tests for resource.go and utils.go. It also adds the capability to add role assignments to Folder.

E2E Results:  
Screenshot 2025-11-14 at 11 09 06

Required ENV:  
TF_ACC=1  
TF_ACC_REGION=eu01  
TF_ACC_ORGANIZATION_ID=xxxx  
STACKIT_SERVICE_ACCOUNT_TOKEN=ey..  
TF_ACC_TEST_PROJECT_SERVICE_ACCOUNT_EMAIL=terraform-xxxx@sa.stackit.cloud

Manual Tests:
```hcl
locals {
  org_id      = "xxxx"
  owner_email = "terraform-xxxx@sa.stackit.cloud"
}

resource "stackit_resourcemanager_folder" "folder" {
  parent_container_id = local.org_id
  name                = "e2e-test-folder"
  owner_email         = local.owner_email
}

resource "stackit_resourcemanager_project" "project" {
  parent_container_id = local.org_id
  name                = "e2e-test-folder"
  owner_email         = local.owner_email
}

resource "stackit_authorization_folder_role_assignment" "fra" {
  resource_id = stackit_resourcemanager_folder.folder.folder_id
  role        = "editor"
  subject     = local.owner_email
}

resource "stackit_authorization_folder_role_assignment" "fra_duplicate" {
  resource_id = stackit_resourcemanager_folder.folder.folder_id
  role        = "editor"
  subject     = local.owner_email
}

resource "stackit_authorization_project_role_assignment" "pra" {
  resource_id = stackit_resourcemanager_project.project.project_id
  role        = "reader"
  subject     = local.owner_email
}

resource "stackit_authorization_organization_role_assignment" "ora" {
  resource_id = local.org_id
  role        = "iaas.project.admin"
  subject     = local.owner_email
}
```

## Checklist

- [x] Issue was linked above
- [x] Code format was applied: `make fmt`
- [x] Examples were added / adjusted (see `examples/` directory)
- [x] Docs are up-to-date: `make generate-docs` (will be checked by CI)
- [x] Unit tests got implemented or updated
- [x] Acceptance tests got implemented or updated (see e.g. [here](https://github.com/stackitcloud/terraform-provider-stackit/blob/f5f99d170996b208672ae684b6da53420e369563/stackit/internal/services/dns/dns_acc_test.go))
- [x] Unit tests are passing: `make test` (will be checked by CI)
- [x] No linter issues: `make lint` (will be checked by CI)

### CODE REVIEW

File: stackit/internal/services/authorization/roleassignments/resource.go [hunk 1]
```diff
@@ -290,69 +327,6 @@ func mapListMembersResponse(resp *authorization.ListMembersResponse, model *Mode
...
-	resourceId := model.ResourceId.ValueString()
-	ctx = tflog.SetField(ctx, "resource_id", resourceId)
-	ctx = tflog.SetField(ctx, "subject", model.Subject.ValueString())
-	ctx = tflog.SetField(ctx, "role", model.Role.ValueString())
-	ctx = tflog.SetField(ctx, "resource_type", r.apiName)
-	return ctx
-}
-
-// returns an error if duplicate role assignment exists
-func (r *roleAssignmentResource) checkDuplicate(ctx context.Context, model Model) error { //nolint:gocritic // A read only copy is required since an api response is parsed into the model and this check should not affect the model parameter
```
Comment (h3adex) [inline stackit/internal/services/authorization/roleassignments/resource.go:354]: Removed this Func. We don't need to check if a duplicate role assignment exists. This logic is handled by the API itself.

File: stackit/internal/services/authorization/testdata/resource-folder-role-assignment.tf [hunk 2]
```diff
@@ -0,0 +1,17 @@
+variable "name" {}
+variable "role" {}
+variable "owner_email" {}
+variable "subject" {}
```
Comment (GokceGK) [inline stackit/internal/services/authorization/testdata/resource-folder-role-assignment.tf:4]: I think we do not need this variable here, or we can use it in _subject = var.subject_ instead of owner_email

File: stackit/internal/services/authorization/testdata/resource-project-role-assignment.tf [hunk 3]
```diff
@@ -0,0 +1,17 @@
+variable "name" {}
+variable "role" {}
+variable "owner_email" {}
+variable "subject" {}
```
Comment (GokceGK) [inline stackit/internal/services/authorization/testdata/resource-project-role-assignment.tf:4]: same here

File: stackit/internal/services/authorization/roleassignments/resource.go (hunk 1 above)
Comment (GokceGK) [reply stackit/internal/services/authorization/roleassignments/resource.go:354]: Having a duplicated role assignment is erroneous. When I use the above given manual tests, the resources are created correctly without any issues, even with duplicated role assignment. The problem is, if we remove **one** of the duplicated role, it removes the role from the resource, and puts the terraform in a weird state. Planning always fails. 

I would have a duplicate check but directly in the planning step, to prevent the initial creation to have duplicated role assignments.

Comment (h3adex) [reply stackit/internal/services/authorization/roleassignments/resource.go:354]: Very good catch. I reintroduced the function in a slightly different way so we can also write unit tests for it without mocking the entire auth client. I also added a test resource to ensure that attempting to create duplicate role assignments on a folder, organization, or project results in an error.

E2E Tests:
Screenshot 2026-01-21 at 13 38 21

File: stackit/internal/services/authorization/roleassignments/resource.go [hunk 4]
```diff
@@ -338,3 +350,18 @@ func mapListMembersResponse(resp *authorization.ListMembersResponse, model *Mode
 
 	return errRoleAssignmentNotFound
 }
+
+// Prevent creating duplicate <resource_id, role, subject> assignments.
```
Comment (h3adex) [inline stackit/internal/services/authorization/roleassignments/resource.go:378]: Wrote a comment above the function to keep everyone aware working on this resource

File: stackit/internal/services/authorization/roleassignments/resource.go [hunk 5]
```diff
@@ -153,39 +151,76 @@ func (r *roleAssignmentResource) Create(ctx context.Context, req resource.Create
...
 	ctx = core.LogResponse(ctx)
 
-	// Map response body to schema
-	err = mapMembersResponse(createResp, &model)
+	err = mapListMembersResponse(listMembersResponse, &model)
 	if err != nil {
+		if errors.Is(err, errRoleAssignmentNotFound) {
+			resp.State.RemoveResource(ctx)
+			return
+		}
```
Comment (marceljk) [inline stackit/internal/services/authorization/roleassignments/resource.go]: This check must be moved to the Read function
